### PR TITLE
fix(bedrock): normalize tool_use ids and names to match Bedrock's regex

### PR DESCRIPTION
**Pull request by Bowl42**
## Summary
- Some clients send tool identifiers like `functions.foo:0` which Bedrock rejects: ``tool_use.id: String should match pattern '^[a-zA-Z0-9_-]+$'``
- Add `NormalizeToolIdentifiers` to substitute disallowed chars with `_` across `tool_use.id`, `tool_result.tool_use_id`, `tool_use.name`, and `tools[].name`
- Wire into `SanitizeForBedrockCompat` so both direct-Bedrock and custom-adapter bedrock-disguise paths pick it up

## Why
Reviewing last few days of FAILED requests on the gateway:
- **23** `tool_use.id` pattern rejections
- **2** `tools[].name` pattern rejections

Substitution is deterministic, so `tool_use.id` and the matching `tool_result.tool_use_id` collapse to the same value without needing a remap. This is a pure compatibility cleanup — no semantic change.

## Test plan
- [x] Unit tests for tool_use.id + matching tool_result.tool_use_id stay paired
- [x] Already-valid identifiers untouched
- [x] tools[].name normalized
- [x] Name capped at 128 chars
- [x] Integration test via `SanitizeForBedrockCompat`
- [ ] CI green

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## 发布说明

* **新增功能**
  * 强化与 Bedrock 的兼容性：自动规范化工具标识符与名称（只保留允许字符、将不合法字符替换、长度上限 128），保证规范化后相关使用与结果标识符保持一致且幂等，冲突时自动消解以避免重复。

* **测试**
  * 新增全面测试，覆盖标识符重写、名称截断、冲突解消、幂等性及与兼容化流程的集成验证。

[Review Change Stack](https://app.coderabbit.ai/change-stack/awsl-project/maxx/pull/557)

### CODE REVIEW

**Comment by Bowl42** [comment]
@coderabbitai please re-review the latest commit (43795d1) which addresses the collision concern with a stable mapping and adds tests.

**Comment by Bowl42** [comment]
@coderabbitai please re-review commit 0892e27, which addresses the regex-hoisting nit and adds 3 more edge-case tests (suffix-skip, repeated originals, empty id).

**Comment by awsldev** [comment]
Blocking review note from local verification.

The `tool_use.id` / `tool_result.tool_use_id` collision issue looks addressed, but I think the same collision class still exists for tool names.

Current behavior normalizes `tools[].name` and `messages[].content[].tool_use.name` by replacing invalid characters with `_`, but it does not disambiguate collisions or share a request-wide name mapping between tool definitions and tool-use blocks.

Example collision:

- `functions.foo`
- `functions/foo`

Both normalize to:

- `functions_foo`

That can collapse two distinct tools into the same Bedrock-visible name, changing tool semantics even though the request passes Bedrock's regex.

I verified this against the current PR head (`0892e27`) with a temporary regression test:

```bash
go test ./internal/adapter/provider/bedrock -run TestPR557ToolNameCollision -count=1
```

Failure shape:

```text
normalized names collide: tools=("functions_foo","functions_foo"), message names=("functions_foo","functions_foo")
```

Minimal failing payload:

```json
{
  "tools": [
    {"name":"functions.foo","description":"a","input_schema":{}},
    {"name":"functions/foo","description":"b","input_schema":{}}
  ],
  "messages": [{"role":"assistant","content":[
    {"type":"tool_use","id":"id1","name":"functions.foo","input":{}},
    {"type":"tool_use","id":"id2","name":"functions/foo","input":{}}
  ]}]
}
```

Suggested fix:

- Add request-wide raw-name -> normalized-name mapping for tool names too.
- Use that same map for both `tools[].name` and `tool_use.name` so definitions and uses stay paired.
- Apply collision disambiguation for names just like ids, while preserving the 128-char cap.
- Add a regression test for `functions.foo` vs `functions/foo` covering both `tools[].name` and `tool_use.name`.

Until that is fixed, I would not merge this PR despite the current green checks.

**Comment by Bowl42** [comment]
@awsl233777 Thanks for the catch — the same collision class did apply to names. Pushed 57ac8d9:

- Factored `newToolIdentifierResolver(maxLen)` so id (no cap) and name (128 cap) share the collision-handling logic.
- Build a request-wide name map shared between `tools[].name` and `tool_use.name`. tools[] is walked first in the first pass so definitions own the canonical slots; tool_use.name references resolve through the same map.
- Suffix allocation re-truncates the base when needed so `base + "_<n>"` stays within 128.
- Added your repro as a regression test (`disambiguates_colliding_tool_names_across_tools[]_and_tool_use`) plus a 128-char-cap suffix test.

File: internal/adapter/provider/bedrock/sanitizer_test.go [hunk 1]
```diff
@@ -488,3 +489,314 @@ func TestSanitizeRequestBodyRemovesModelAndStream(t *testing.T) {
...
+					{"type":"tool_use","id":"","name":"a","input":{}}
+				]}
+			]
+		}`)
+
+		result := NormalizeToolIdentifiers(body)
+		if got := gjson.GetBytes(result, "messages.0.content.0.id").String(); got != "" {
+			t.Errorf("empty id should stay empty, got %q", got)
+		}
+	})
```
[the remark this answers is not in the document]
Comment (Bowl42) [reply internal/adapter/provider/bedrock/sanitizer_test.go]: Good catch on the spec citation. I'm taking option (c): drop the test rather than synthesize.

Two empty-id originals can't reliably pair across tool_use ↔ tool_result (the original signal is gone), so synthesizing would either collide (multiple empties → same synthetic id) or invent unstable values. The implementation's empty short-circuit is incidental — leaving the malformed request to surface Bedrock's exact validation error is more honest than blessing a passthrough as a contract. Pushed f4a4340 removing the misleading test.

Length cap (64 chars for Converse, unbounded for InvokeModel/Anthropic format) isn't touched: the observed production failures all came through the Anthropic-format relay path which doesn't cap, and a 64-cap on InvokeModel ids could truncate valid long ids and break pair-references for clients that use them.

**Review by awsl233777 — APPROVED** [review]
Approved after checks passed and CodeRabbit reported no actionable comments on the latest revision.
